File: frontend/src/types/task.ts
import type { PaginationParams } from "@/types/pagination";

export const taskPriorities = ["low", "medium", "high", "urgent"] as const;
export const taskStatuses = ["open", "in_progress", "completed", "cancelled"] as const;

export type TaskPriority = (typeof taskPriorities)[number];
export type TaskStatus = (typeof taskStatuses)[number];

export type Task = {
  assigned_user_id: string;
  created_at: string;
  description: string | null;
  due_date: string | null;
  id: string;
  organization_id: string;
  priority: TaskPriority;
  status: TaskStatus;
  title: string;
  updated_at: string;
};

export type TaskInput = {
  assigned_user_id?: string | null;
  description: string | null;
  due_date: string | null;
  priority: TaskPriority;
  status: TaskStatus;
  title: string;
};

export type TasksListParams = PaginationParams & {
  assigned_user_id?: string;
  search?: string;
  sort_by?: "created_at" | "due_date" | "priority";
  sort_direction?: "asc" | "desc";
  status?: TaskStatus;
};

File: backend/app/schemas/task.py
from datetime import datetime
from typing import Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, StringConstraints, field_validator

from app.schemas.common import RequestSchema

TaskTitle = Annotated[str, StringConstraints(min_length=1, max_length=255)]
TaskDescription = Annotated[str, StringConstraints(max_length=20_000)]
TaskPriority = Literal["low", "medium", "high", "urgent"]
TaskStatus = Literal["open", "in_progress", "completed", "cancelled"]


class TaskCreate(RequestSchema):
    title: TaskTitle
    description: TaskDescription | None = None
    priority: TaskPriority = "medium"
    status: TaskStatus = "open"
    due_date: datetime | None = None
    assigned_user_id: UUID | None = None


class TaskUpdate(RequestSchema):
    title: TaskTitle | None = None
    description: TaskDescription | None = None
    priority: TaskPriority | None = None
    status: TaskStatus | None = None
    due_date: datetime | None = None
    assigned_user_id: UUID | None = None

    @field_validator("title", "priority", "status", "assigned_user_id")
    @classmethod
    def require_non_null_required_fields(cls, value: object) -> object:
        if value is None:
            raise ValueError("This field cannot be null")
        return value


class TaskResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    assigned_user_id: UUID
    title: str
    description: str | None
    priority: TaskPriority
    status: TaskStatus
    due_date: datetime | None
    created_at: datetime
    updated_at: datetime
